### PR TITLE
Make Colon immutable

### PR DESCRIPTION
**Pull request by tkoolen**
Fixes https://github.com/JuliaLang/julia/issues/18034.
Fixes https://github.com/andyferris/StaticArrays.jl/issues/17.

### CODE REVIEW

**Comment by timholy** [comment]
An exemplary first PR. You look like a pro already.

**Comment by tkoolen** [comment]
Thank you.
Looks like the OSX build exceeded a time limit.

File: test/subarray.jl [hunk 1]
```diff
@@ -484,3 +484,15 @@ u = (1,2:3)
...
     @test X[1:end,2,2] == @view X[1:end,2,2]
 end
+
+# issue #18034
+# ensure that it is possible to create an isbits, LinearFast view of an immutable Array
+let
+    immutable ImmutableTestArray{T, N} <: Base.DenseArray{T, N}
+    end
+    Base.size(::Union{ImmutableTestArray, Type{ImmutableTestArray}}) = (0, 0)
+    Base.linearindexing(::Union{ImmutableTestArray, Type{ImmutableTestArray}}) = Base.LinearFast()
```
Comment (musm) [inline test/subarray.jl]: Wouldn't the following suffice. I don't' see the reason for the Unions?
```julia
Base.size(I::ImmutableTestArray) = (0,0)
Base.linearindexing{T<:ImmutableTestArray}(::Type{T}) = Base.LinearFast()
```

Comment (tkoolen) [inline test/subarray.jl]: True. I basically just copied and slightly modified these methods from `StaticArrays.SMatrix`and`StaticArrays.StaticArray`.

**Pull request merged**